pkg/controller/servicednsendpoint/servicednsendpoint_controller.go

```diff
@@ -0,0 +1,464 @@
+		return true
+	}
+	if !reflect.DeepEqual(oldObject.Status, newObject.Status) {
+		return true
+	}
+
+	return false
+}
+
+// obj could be an *feddnsv1a1.MultiClusterServiceDNSRecord, or a DeletionFinalStateUnknown marker item.
+func (d *ServiceDNSEndpointController) enqueueObject(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %#v: %v", obj, err)
+		return
+	}
+	d.queue.Add(key)
+}
+
+func (d *ServiceDNSEndpointController) worker() {
+	// processNextWorkItem will automatically wait until there's work available
+	for d.processNextItem() {
+		// continue looping
+	}
+}
+
+func (d *ServiceDNSEndpointController) processNextItem() bool {
+	key, quit := d.queue.Get()
+	if quit {
+		return false
+	}
+	defer d.queue.Done(key)
+
+	err := d.processItem(key.(string))
+
+	if err == nil {
+		// No error, tell the queue to stop tracking history
+		d.queue.Forget(key)
+	} else if d.queue.NumRequeues(key) < maxRetries {
+		glog.Errorf("Error processing %s (will retry): %v", key, err)
+		// requeue the item to work on later
+		d.queue.AddRateLimited(key)
+	} else {
+		// err != nil and too many retries
+		glog.Errorf("Error processing %s (giving up): %v", key, err)
+		d.queue.Forget(key)
+		runtime.HandleError(err)
+	}
+
+	return true
+}
+
+func (d *ServiceDNSEndpointController) processItem(key string) error {
+	startTime := time.Now()
+	glog.V(4).Infof("Processing change to ServiceDNSEndpoint %s", key)
+	defer func() {
+		glog.V(4).Infof("Finished processing ServiceDNSEndpoint %q (%v)", key, time.Since(startTime))
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	obj, exists, err := d.serviceDNSObjectStore.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("error fetching object with key %s from store: %v", key, err)
+	}
+
+	if !exists {
+		//delete corresponding DNSEndpoint object
+		return d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Delete(name, &metav1.DeleteOptions{})
+	}
+
+	dnsObject, ok := obj.(*feddnsv1a1.MultiClusterServiceDNSRecord)
+	if !ok {
+		return fmt.Errorf("recieved event for unknown object %v", obj)
+	}
+
+	dnsEndpoints, err := GetEndpointsForServiceDNSObject(dnsObject)
+	if err != nil {
+		return err
+	}
+
+	dnsEndpointObject, err := d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			newDNSEndpointObject := &feddnsv1a1.DNSEndpoint{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      name,
+					Namespace: namespace,
+				},
+				Spec: feddnsv1a1.DNSEndpointSpec{
+					Endpoints: dnsEndpoints,
+				},
+			}
+
+			_, err = d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Create(newDNSEndpointObject)
+		}
+		return err
+	}
+
+	dnsEndpointObject.Spec.Endpoints = dnsEndpoints
+	_, err = d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Update(dnsEndpointObject)
```

Comment:
Does it make sense to update only if there is a difference in existing and newly computed endpoints?

Reply:
good catch. missed it earlier. updated now.